### DISCUSSION TITLE
Custom validator with arguments

### DISCUSSION THREAD

**Discussion by galaczi** [adonisjs/core#3651] [Help] [answered]
I would like to add a custom rule to filter out bad words from an input. It would take an array of strings just like the ``notIn`` rule. The documentation shows an example where no argument is needed. It is not clear what is the ``options`` objects since it does not contain the arguments I pass in a rule. How would I get those arguments inside the ``validator.rule()`` callback?

**Comment by galaczi** [comment]
I had to dig into the source code a bit to find that the second argument in the callback function should be ``compiledOptions``. This is what I came up with:

```
validator.rule('doesNotContain', (value: string, compiledOptions: string[][], options) => {
  const found = compiledOptions.find((option) => value.includes(option))

  if (found) {
    options.errorReporter.report(
      options.pointer,
      'doesNotContain',
      `${found} is not allowed`,
      options.arrayExpressionPointer
    )
  }

  return
})
```

I still have questions.

1. What are the pointers in the ``options`` object?
2. Can I somehow return the ``found`` variable so I can access it for custom validation messages?
3. If I return out of the function does that mean the validation passed?

**Comment by tomgobich** [comment, marked as answer]
I'm a bit confused why you're not just using the `notIn` rule? It looks like you're pretty much just recreating it under a different name. 

Also, it sounds like the below is the portion of the documentation you were looking for. It shows how to pass arguments into a custom validation rule and it shows how to use your custom rule's error properties to provide a custom message.
https://docs.adonisjs.com/guides/validator/custom-rules#passing-options-to-the-rule

You can also find more information about the error reporting system and what pointers are here:
https://docs.adonisjs.com/guides/validator/error-reporters#creating-your-error-reporter

The validation is passed if no errors are reported using the `errorReporter` and if no exceptions are thrown. Returning merely stop execution of further code within your rule.

**Reply by galaczi** [reply]
``notIn`` validates the value to ensure it is not inside an array of provided values. However I am doing the opposite. Validate that the value (from a textarea) does not contain any of the array items (passed in to the rule). 

Thanks for the additional explanation and the links! I don't know how I missed that part.

**Reply by tomgobich** [reply]
Oh, gotcha! That makes sense then, lol sorry.

Anytime!
